### ISSUE TITLE
Add custom projection mode to Camera3D, make Camera3D store its CameraMatrix and expose CameraMatrix to user code

### ISSUE THREAD

**Issue by Astrono2** [godotengine/godot-proposals#2713]
### Describe the project you are working on
I'm working on a portal plugin.

### Describe the problem or limitation you are having in your project
I need to use oblique frustum projection to allow for the portals' cameras to clip content that's behind the portal plane. Which is currently impossible in Godot.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation
The enhancement would be to add a new projection mode, PROJECTION_CUSTOM to the Camera class, which would allow the developer to directly set the matrix to be used for projection. For my specific case, that would allow me to use oblique frustum projection. It would also address any other projection that any developer could want to use. And it would probably allow for some weird distortion effects as well.
Also, if a new projection is added with a CameraMatrix exposed to the user, it would make sense to migrate the entire Camera class and camera rendering to work with a CameraMatrix object belonging to each Camera instead of calculating the matrices from the Cameras' parameters.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams
I think this can be done in a few steps:

1. Make CameraMatrix or other 4x4 matrix type available to user code
    1. Register CameraMatrix so that it is accessible to user code
    2. Add CameraMatrix to the documentation
2. Change the Camera class to allow for custom matrices
    1. Add a CameraMatrix object to the Camera class (could be named projection_matrix)
    2. Add updating the matrix when a parameter is changed (fov, near, far, etc.) by modifying the appropriate item in the matrix
    3. Add updating the matrix when the projection mode is changed by using the already existing methods in the CameraMatrix struct
    4. Add PROJECTION_CUSTOM to the Projection enum
    5. Add the CameraMatrix to the Camera's properties
3. Replace steps like [this one](https://github.com/godotengine/godot/blob/e2a2138d2bcd5f5f5c7dde8d9371c62192d459f1/servers/rendering/renderer_scene_cull.cpp#L2439) in the rendering process with getting the camera's projection matrix.
4. Add necessary methods to either Camera3D or CameraMatrix to replace the functionality that calculating the matrix when necessary provided (e.g. setting the aspect ratio). So that the matrix doesn't have to be completely remade
5. Update Camera3D and possibly RenderingServer documentation
6. XR wizardry I cannot comprehend.

Here's a little mockup:
![image](https://user-images.githubusercontent.com/8162008/117874349-3d0c4a00-b277-11eb-8b8d-1d7a8d508697.png) ![image](https://user-images.githubusercontent.com/8162008/117874394-4ac1cf80-b277-11eb-9ad7-f65296df161c.png)
![image](https://user-images.githubusercontent.com/8162008/125628372-33481318-2111-48ad-a33c-6a9d8ea1fefa.png)

### If this enhancement will not be used often, can it be worked around with a few lines of script?
Most definitely not

### Is there a reason why this should be core and not an add-on in the asset library?
It requires considerable (although not very invasive) changes to two core classes and the rendering system.

**Comment by lyuma** [comment]
Related to godotengine/godot#7499

My understanding is this feature is used for creating oblique frustums (mirrors / portals). Would a more limited feature like PROJECTION_OBLIQUE be sufficient to cover your proposed usecases?

**Comment by SaracenOne** [comment]
Yes, our specific use-case for this is creating portals/mirrors, particularly in VR. I would add also that the ability to seperate the view and clipping frustum seperately would likely go further in allowing optimisation, because we could limit the clipping plane to the confines of the portal.

**Comment by io12** [comment]
I was also working on a [portal plugin](https://github.com/io12/godot-portal-demo) and needed this (in addition to other things, like stencil buffering).

**Comment by elvisish** [comment]
> I was also working on a [portal plugin](https://github.com/io12/godot-portal-demo) and needed this (in addition to other things, like stencil buffering).

It's one of the few things I've missed about switching to godot, having full access to the camera projection matrix. This portal plugin looks like it could be amazing if it could be completed with oblique frustum projection being available.

**Comment by Astrono2** [comment]
> Related to [godotengine/godot#7499](https://github.com/godotengine/godot/issues/7499)
> 
> My understanding is this feature is used for creating oblique frustums (mirrors / portals). Would a more limited feature like PROJECTION_OBLIQUE be sufficient to cover your proposed usecases?

That's also the only use case I know, but in general I think having access to a Camera's matrix is a better solution. Specially since the changes I propose would also simplify adding new projection modes in the future. At the moment, the [VisualServerScene::render_camera](https://github.com/godotengine/godot/blob/39826d3a942b482d7d25bb4d9882a8436cb0e8e9/servers/visual/visual_server_scene.cpp#L1811) function scales with the amount of projection modes, as well as many of the functions in [Camera](https://github.com/godotengine/godot/blob/39826d3a942b482d7d25bb4d9882a8436cb0e8e9/scene/3d/camera.cpp).
```cpp
CameraMatrix cm;

if (mode == PROJECTION_ORTHOGONAL) {
	cm.set_orthogonal(size, viewport_size.aspect(), p_z_depth, far, keep_aspect == KEEP_WIDTH);
} else {
	cm.set_perspective(fov, viewport_size.aspect(), p_z_depth, far, keep_aspect == KEEP_WIDTH);
}
```
This appears 3 times in camera.cpp. Storing the Camera's matrix in a variable would make the code a lot simpler.

**Comment by winston-yallow** [comment]
Exposing the possibility to set the matrix directly would be really nice in my opinion. It will also allow for more artistic effects like morphing between two matrices (even if not perspective/mathematically correct during morph you can achieve quite some interesting effects).

**Comment by wareya** [comment]
I want this to make it easier to fake 2d-style isometric and orthographic perspectives using 3d assets. Right now I need to stretch the world in bizarre ways to get the same effect, and it loses me a lot of control.

**Comment by Astrono2** [comment]
I left this hanging for a long time due to school. So I just edited to comply with not touching the Variant class and to update it for Godot 4, and I'll start working on it now!

**Comment by Calinou** [comment]
> I left this hanging for a long time due to school. So I just edited to comply with not touching the Variant class and to update it for Godot 4, and I'll start working on it now!

Bump :slightly_smiling_face: Have you made progress so far?

**Comment by Astrono2** [comment]
> > I left this hanging for a long time due to school. So I just edited to comply with not touching the Variant class and to update it for Godot 4, and I'll start working on it now!
> 
> Bump slightly_smiling_face Have you made progress so far?

Sadly, no. I really want to work on this, but school is hitting hard. I think I'll get to it once I graduate at the end of the year. I have been reading the code and making minor changes to see what effects modifying this or that has, but not much more.

**Comment by qaptoR** [comment]
Hey all, I've just successfully implemented and Oblique Near Plane projection camera via the Godot source code.

Technically, I have Igor Kosyanenko to thank for doing the heavy lifting, specifically figuring out how to apply Eric Lengyel's near clipping matrix transformation algorithm.

Igor originally posted his project status here:
https://godotengine.org/qa/64543/oblique-near-clipping-plane-for-camera?show=120466#c120466

And Eric's algorithm is found here (also linked in Igor's post):
http://terathon.com/code/oblique.html

None of Igor's project or source code is available anymore, so I had to make the adjustments to the engine (specifically: additions) adding the new camera type.

I have done my best to make the additions as minimal as possible (pretty much get the engine to compile, and render the proper oblique near plane).
I'm not sure yet what the appropriate channels are to go through to make a pull request (if it's even necessary), but I think it's at least ready for anyone looking for this type of camera to try it out in their projects.

The code can be found on my forked godot repo on this branch:
https://github.com/qaptoR/godot/tree/3.4_oblique

let me know what you think!


Lastly: if you don't check out Igor's post, is that he specifically got stuck because his skybox was warping when seen through the portals camera view. The solution turned out to be setting the custom FOV for the skybox of whatever environment the portal camera can see. 
The FOV should be the same as the portal camera (which should be the same as the player camera).

This is a short video showing the problem, and the solution:
https://youtu.be/hchttF-iN7Y

**Comment by qaptoR** [comment]
Hey. So, I just realized I dropped the code, but no explanation of the interface provided for the new camera. So here are some pictures:

This is the new option in the drop-down list
![Oblique Camera Selection](https://user-images.githubusercontent.com/40372043/141320865-d82d9192-199e-4a08-a63d-bab2f9f791bd.png)

This is the new property, that can be manipulated in the 3d view editor
(thanks to Godot's great modularity)
![Oblique Plane property](https://user-images.githubusercontent.com/40372043/141320944-9c81d494-cc5f-4bed-aeee-6158aa88ca82.png)

This is the code attached to the portal's camera. The most relevant thing here is to make sure that the 'clip_plane' variable is the global_transform of the portal's mesh with the viewport_texture
![Code](https://user-images.githubusercontent.com/40372043/141321177-8a736dff-ac97-4da4-995e-c0fecce36ff2.png)

These are some functions provided by Igor, that essentially replace the unity functions used by Sebastian Lague in his portal tutorial
![support code](https://user-images.githubusercontent.com/40372043/141321560-14a0df2f-0416-4dd3-a6b6-4cf166a5daf5.png)

Have fun!

**Comment by elvisish** [comment]
> Hey all, I've just successfully implemented an Oblique Near Plane projection camera via the Godot source code.
> 
> Technically, I have Igor Kosyanenko to thank for doing the heavy lifting, specifically: figuring out how to apply Eric Lengyel's near clipping matrix transformation algorithm.
> 
> Igor originally posted his project status here: https://godotengine.org/qa/64543/oblique-near-clipping-plane-for-camera?show=120466#c120466
> 
> And Eric's algorithm is found here (also linked in Igor's post): http://terathon.com/code/oblique.html
> 
> None of Igor's project or source code is available anymore, so I had to make the adjustments to the engine (specifically: additions) adding the new camera type.
> 
> I have done my best to make the additions as minimal as possible (pretty much get the engine to compile, and render the proper oblique near plane). I'm not sure yet what the appropriate channels are to go through to make a pull request (if it's even necessary), but I think it's at least ready for anyone looking for this type of camera to try it out in their projects.
> 
> The code can be found on my forked Godot repo on this branch: https://github.com/qaptoR/godot/tree/3.4_oblique
> 
> let me know what you think!
> 
> Lastly, if you don't check out Igor's post: he specifically got stuck because his skybox was warping when seen through the portals camera view. The solution turned out to be setting a custom FOV for the skybox of whatever environment the portal camera can see. The FOV should be the same as the portal camera (which should be the same as the player camera).
> 
> This is a short video showing the problem, and the solution: https://youtu.be/hchttF-iN7Y

This is really amazing, I'll try it later, thanks so much for all of your work!

**Comment by qaptoR** [comment]
Thanks! I'm really grateful for everyone who took part in the discussion about this camera type over the years (which I came across in my research). All the back and forth between everyone really pointed me in the right direction of how to implement it.

This comment in particular pointed me to the right part of the engine to put the additions:
https://github.com/godotengine/godot-proposals/issues/501#issuecomment-725030633


It really just goes to show how important documentation is. If anyone along the way hadn't bothered to post their progress, there's no telling where I'd be in the process right now.

**Comment by qaptoR** [comment]
I realized that the line: projection = PROJECTION_OBLIQUE
is redundant. It was a holdout from Igor's interpretation of Lague's approach. They switch back to a perspective camera when the portal camera in within a certain threshold... though I'm not sure why, leaving it as an oblique camera doesn't cause any issues.
![image](https://user-images.githubusercontent.com/40372043/141372935-fd4c5292-63bc-4121-a7df-b897bf036f56.png)

Though they use two variables 'near_clip_offset' and 'near_clip_limit', the latter of which determines when the camera switches back to perspective (cam_space_dst <= near_clip_limit). 
I didn't play around much with the near_clip_limit, but I did with the near_clip_offset: there's an issue when looking at the portals from extremely obtuse angles, where the camera can see where clipping plane cuts of the geometry.

I rationalized that this is why most portals are oval in shape, since the issue presents most prominently in the corners, and using a mask to cut them off would solve it.

I specifically want my portals to be rectangular, because I want them to give off that 'time-door' vibe from LOKI. So I applied a rectangular mask to the portal viewport mesh texture, and it nearly solves the problem, but since it's not ovoid, the corners still cause an issue. The second trick I've applied is deepening the frame around my portals, so that at extremely obtuse angles it's nearly impossible to see behind the clipped edges (except when I literally try to force drastically obtuse looking angles, which are unlikely to happen often/naturally, so at this point, I'm gucci w/ it.

**Comment by Astrono2** [comment]
> Hey all, I've just successfully implemented an Oblique Near Plane projection camera via the Godot source code.
> 
> Technically, I have Igor Kosyanenko to thank for doing the heavy lifting, specifically: figuring out how to apply Eric Lengyel's near clipping matrix transformation algorithm.
> 
> Igor originally posted his project status here: https://godotengine.org/qa/64543/oblique-near-clipping-plane-for-camera?show=120466#c120466
> 
> And Eric's algorithm is found here (also linked in Igor's post): http://terathon.com/code/oblique.html
> 
> None of Igor's project or source code is available anymore, so I had to make the adjustments to the engine (specifically: additions) adding the new camera type.
> 
> I have done my best to make the additions as minimal as possible (pretty much get the engine to compile, and render the proper oblique near plane). I'm not sure yet what the appropriate channels are to go through to make a pull request (if it's even necessary), but I think it's at least ready for anyone looking for this type of camera to try it out in their projects.
> 
> The code can be found on my forked Godot repo on this branch: https://github.com/qaptoR/godot/tree/3.4_oblique
> 
> let me know what you think!
> 
> Lastly, if you don't check out Igor's post: he specifically got stuck because his skybox was warping when seen through the portals camera view. The solution turned out to be setting a custom FOV for the skybox of whatever environment the portal camera can see. The FOV should be the same as the portal camera (which should be the same as the player camera).
> 
> This is a short video showing the problem, and the solution: https://youtu.be/hchttF-iN7Y

This is amazing! I still think giving the user access to the camera matrix is a good idea. But this covers my use case. I'll check it out when I have the time to get back to this!

**Comment by qaptoR** [comment]
Actually, now that you mention it... 

I think it would be a good idea to transition the gdscript code for finding the clipping plane in camera space to the camera itself. That way, it would be easy to pass the camera the origin and normal of any plane, and get an oblique near clipping plane from that, rather than needing to attach a script every time (though you would still need a script to continuously update the values)

While I'm at it.. I'll make an attempt at exposing the CameraMatrix...
I think that will be slightly more challenging, since it would make sense to expose the whole CameraMatrix class to the editor, enabling us to directly access matrix manipulation functions without wrapping them into the Camera class.

Thoughts?

**Comment by qaptoR** [comment]
Okay, so I've moved the camera-space math into the set_oblique_plane function, and it works exactly as before (probably better performance).

Here's the code changed:
![image](https://user-images.githubusercontent.com/40372043/141523098-71c926f5-7058-4f14-b935-ba79e566dfcd.png)

and what it looks like in the portal_camera script:
![image](https://user-images.githubusercontent.com/40372043/141523180-4c962b44-0c43-45f3-bee4-84ce4966175e.png)

much, cleaner now; thank you for sparking the idea in my head to move the code for a sleeker interface.

I think I'm going to hold off on exposing the matrix to the editor, since there are multiple ways to implement it... and I'm not sure which would be best:
expose the CameraMatrix class directly, create a new MatrixCamera node with an extended api, or simply extend the Camera api, both latter options wrapping CameraMatrix api functions to work directly on it's matrix (and figure out how to represent the 4x4 matrix in the editor).

**Comment by qaptoR** [comment]
Hey, I know I'm posting a lot here, I just know most of you are interested in implementing portals, which is my main reason for making additions to the engine (I need them too).

turns out that multiple setters was.. working somehow? but apparently also broken? I believe I may have just convinced myself it worked. In any case, godot does not like more than 1 parameter for setters, so I switched to a dictionary format for the oblique camera.

In addition to that, I've just finished implementing a new feature to the spatialmaterial, which will allow the specification of an arbitrary world plane to clip a mesh with.

Initially it was just a special shader code I found from a video:
https://github.com/godotengine/godot/issues/3499#issuecomment-602129675

and it worked well for basic uses, but I realized that when importing models with attached materials, I would have to convert those materials into shaders, and then inject the clipping code manually.. for every material on the model, and for every model (and that just won't cut it for me).

So after more deliberation than I care to admit (trying to figure out if I could use a 'next pass' shader to accomplish the same effect) I resolved to make an addition to the spatialmaterial, which can be toggled on and off. 

The code for it is on the same branch as before, but here's a video showing both features in more detail (probably best to watch at 2x speed, i'm talking slower than it felt like).

https://youtu.be/zT6Zlx09UQY

**Comment by Astrono2** [comment]
I don't think modifying `SpatialMaterial` for this is a good idea. I toyed with the exact same idea, but my approach was to convert the spatial material to a shader and then inject the code. This approach doesn't require modifying the engine. (Also don't forget that engine changes should address specific things and make changes only for implementing those). This proposal is about giving users more control of the camera, any changes you want to make to `SpatialMaterial` should have their own proposal.

**Comment by qaptoR** [comment]
Sorry, I'm new to contributing to open source projects, and I understand that there's an etiquette I'm not yet aware of.

My last post was only to express a necessary fix to the oblique camera I hadn't realized was broken, and an additional change I had made to the engine on the same branch to the spatialmaterial. It was not a feature proposal, but for anyone who might pull from my branch might have noticed and wondered about why it was there (or possibly why it wasn't there in the official release).

I'm presently just starting the process of looking into how to go about setting up a pull-request formally for the oblique camera, there's a lot more documentation involved than I realized, so it won't happen immediately.
In the meantime, I will leave my 3.4_oblique branch as-is without further additions (or at least until I can figure out how to roll-back the spatialmaterial onto a separate branch).

On the spatialmaterial: it addressed my needs precisely, and is very useful for the 'passing through the portal' effect without needing to inject code into every material-made-a-shader manually.

I'm just enthusiastic about how much easier it is to improve the engine for my needs than I expected, but nothing I have presented thus far has been with the expectation that it would be integrated into the upstream repo.

**Comment by winston-yallow** [comment]
@qaptoR since you write that you are new, are you aware of the contributors chat? It's located at https://chat.godotengine.org/

There also is a `#new-contributors` channel over there in case you have questions about the PR workflow or related stuff.

**Comment by Astrono2** [comment]
Wow, okay. I sat down to get to work on this and the rendering system has gotten so much more complex since Godot 3. The handling and passing around of `CameraMatrix` objects is all over the place. This will take an enormous amount of reading code and drawing things out to even grasp what files are involved. I don't think I'll get it done by the release of Godot 4.

Sadly what was a non-trivial but doable change when I proposed it has grown in scope drastically. I think it would be best to close this with @qaptoR's eventual pull request. I still think storing a `Camera3D`'s matrix as a property and using that for rendering is a good idea, but it's no longer a sensible solution to anyone's needs. It would be more of a structural change to allow for easier customization and more user control.

I will open another issue once I have a better idea of what storing a `Camera3D`'s matrix and using it for rendering even means in the current rendering pipeline.

**Comment by qaptoR** [comment]
So I've spent some time the past few days chipping away at getting the PR ready: something has come up that I think I should mention before I go further.

My solution is for version 3.4 because it was the latest stable branch. From the docs it sounds like it's better to target master so the work can be cherry-picked for previous versions. But the work is already done. I had planned to implement it for 4.0 too (for myself) so I don't mind doing that sooner rather than later, but should I finish the PR for 3.X in the meantime?

I'm pretty sure all I have left to do is the reference documentation.

- I've used clang-format on the affected files (and made sure nothing went awry)
- I've cleaned up the commit history (on a new branch in case I made a mistake) so it's a single commit and the SpatialMaterial changes are not included anymore

Please let me know if I've missed anything important.

**Comment by Astrono2** [comment]
As far as I know, you're good to go. Make the pull request and if you missed anything the maintainers will let you know.

**Comment by qaptoR** [comment]
Sorry that PR took so long to complete. I was far more depleted after last semester than I thought, and I pretty much forgot about everything when I realized Halo Infinite had dropped.

For anyone who ends up looking at the PR, please let me know if there's anything I should include in the message (or take out for redundancy).

**Comment by ngreve** [comment]
Looking at your PR, the camera matrix is not fully exposed?
I was looking for a way to tween between the `orthogonal` and `projection` camera mode. Would this be possible with your PR? 
Here is a link to the [Unity forum](https://forum.unity.com/threads/smooth-transition-between-perspective-and-orthographic-modes.32765/), which also explains what this is about.

**Comment by wareya** [comment]
Why *not* just expose the matrix directly?

**Comment by ngreve** [comment]
I just want to make sure this proposal does not get closed then. Even though your implementation does seem to serve some of the use-cases, the title and the whole proposal description is aiming for a more general approach.

> [..] and not necessarily enough use-cases to justify it.

Please take a look at the first comments of this thread:

- [It's one of the few things I've missed about switching to godot, having full access to the camera projection matrix.](https://github.com/godotengine/godot-proposals/issues/2713#issuecomment-840811001)
- [[..] but in general I think having access to a Camera's matrix is a better solution.](https://github.com/godotengine/godot-proposals/issues/2713#issuecomment-841182597)
- [It will also allow for more artistic effects like morphing between two matrices](https://github.com/godotengine/godot-proposals/issues/2713#issuecomment-841250818)
- [Right now I need to stretch the world in bizarre ways to get the same effect, and it loses me a lot of control.](https://github.com/godotengine/godot-proposals/issues/2713#issuecomment-845853518)

I wouldn't be to fast with statements like that. People use stuff, when it is given to them. Secondly, just a small fraction of developers are aware of this proposal, so just a small amount will make themselves heard - even though here are actually quite a few who are saying that they need/want to access the matrix directly . Thirdly, Godot has to compare itself to other engines and what can be done with them (see my earlier post and the link to the Unity forum).

> by having another camera type [..] and then exposing a lerp value between 0 and 1 [..]

I don't want to lean too far out of the window, but that just proves my point, at least somewhat. Currently some contributor would  have to implement a new camera type for each possible creative idea a game dev has, regarding the camera. Concept wise, it seems cleaner and more maintainable to store the camera matrix in the camera object itself and simply expose it to the developer, so they can decide what to do with it.

I don't want to talk you into implementing the more general approach. Again, I just want to prevent this proposal from getting closed by the mods.

**Comment by qaptoR** [comment]
Yes, sorry for my wording, it was misleading. I didn't mean to imply there weren't enough use-cases throughout the community. I meant that it appears to be more difficult for what they might reasonably need it for. Like, in my case when I started researching what I needed to implement the portal effect I learned I would need to update the camera-matrix which is only exposed in the visual_server. So I opted to work within the existing limits to reach my goals.

Then again, when I started I was a lot less familiar with the engine architecture and it was far more daunting.

**Comment by ngreve** [comment]
> to implement the portal effect

I am not sure if even this justifies the implementation of the given camera type. I tried to implement Sebastian's portal tutorial in Godot too and got the basics running by simply using shaders, just like he did in his tutorial.

https://user-images.githubusercontent.com/18660100/148828656-925ae762-5318-4ec9-b1ba-2711f114d577.mp4

Am I missing something?

**Comment by qaptoR** [comment]
try putting an obstacle in front of the portal
The PR solves the issue of obstacles between the portal camera and it's parent portal obstruct the camera's view by setting the camera's near clipping plane in parallel with the portal, culling everything in between.

**Comment by ngreve** [comment]
Of curse, thanks for the clarification.

**Comment by Astrono2** [comment]
> Why _not_ just expose the matrix directly?

That is a better solution, but it would require having the camera store the matrix and use that for the rendering. In 3.x that would be relatively easy but, given the complexity of the rendering server in master, it's a far more involved task. I plan on working on it eventually but it exceeded the scope that I thought it had. I mentioned above that @qaptoR 's PR is closer to addressing the portal use case in a reasonable time than exposing the matrix, even though I still consider exposing it a far better solution.

**Comment by SmartySmart702** [comment]
Any news on this?

**Comment by elvisish** [comment]
For clarification, it's still not possible to do true portals as of March 2022?

**Comment by Calinou** [comment]
> For clarification, it's still not possible to do true portals as of March 2022?

No, not yet.

**Comment by qaptoR** [comment]
> Any news on this?
> For clarification, it's still not possible to do true portals as of March 2022?

My PR here https://github.com/godotengine/godot/pull/56664 contains code for a camera that can do the oblique near plane clipping which is all that is missing from Godot. I recommend learning to compile Godot from source regardless of trying my camera. It has opened up a whole new world of new features that I get to add/test to my own version from other things I do myself as well as PRs from the community. 

Here is a simple demonstration of the camera's effect: 
https://www.youtube.com/watch?v=PipVd6wRnMk

Here is my progress on my own portal mechanic using the camera:
https://www.youtube.com/watch?v=EHC2OUpDzQI

I actually had to back-track and re-assess things since making that last video because even though the instant translation was working, I learned I was doing different calculations at different local/global scopes which just made the whole thing really jank.
I'm slowly learning how to use quaternions and matrix multiplication to transition between frames of reference, and my progress has been incredible. I've just got a few big assignments due soon at Uni. so I might not get back to this until the summer starts. 

In any case, I've just been translating Sebastian Lagues Tutorial on portals, but I plan to make my own that is specific to how It was accomplished in Godot, providing the different shaders that have been translated as well.

Before that happens, I do recommend trying to translate his tutorial yourself as a challenge. I've learned so much about Godot, shaders, quaternions, projection, etc. just from this one mechanic.

**Comment by elvisish** [comment]
> > Any news on this?
> > For clarification, it's still not possible to do true portals as of March 2022?
> 
> My PR here [godotengine/godot#56664](https://github.com/godotengine/godot/pull/56664) contains code for a camera that can do the oblique near plane clipping which is all that is missing from Godot. I recommend learning to compile Godot from source regardless of trying my camera. It has opened up a whole new world of new features that I get to add/test to my own version from other things I do myself as well as PRs from the community.
> 
> Here is a simple demonstration of the camera's effect: https://www.youtube.com/watch?v=PipVd6wRnMk
> 
> Here is my progress on my own portal mechanic using the camera: https://www.youtube.com/watch?v=EHC2OUpDzQI
> 
> I actually had to back-track and re-assess things since making that last video because even though the instant translation was working, I learned I was doing different calculations at different local/global scopes which just made the whole thing really jank. I'm slowly learning how to use quaternions and matrix multiplication to transition between frames of reference, and my progress has been incredible. I've just got a few big assignments due soon at Uni. so I might not get back to this until the summer starts.
> 
> In any case, I've just been translating Sebastian Lagues Tutorial on portals, but I plan to make my own that is specific to how It was accomplished in Godot, providing the different shaders that have been translated as well.
> 
> Before that happens, I do recommend trying to translate his tutorial yourself as a challenge. I've learned so much about Godot, shaders, quaternions, projection, etc. just from this one mechanic.

I added a comment on your video about translating position, rotation and velocity that might be useful, but don't want to divert this too much from the camera matrix discussion ;)

**Comment by stefano-elysium** [comment]
I would also like to be able to modify a camera's projection matrix a la
https://docs.unity3d.com/ScriptReference/Camera-projectionMatrix.html
in order to do cavalier projection, which is not a orthogonal isometric projection.

**Comment by douglaslassance** [comment]
I'm curious to know if [this](https://jsfiddle.net/e28dqzk0/55/) can be achieved using a shader. I am currently branching off the engine code and it seems a bit overkill.

**Comment by douglaslassance** [comment]
Testing in Unity convinced me that having the ability to get and set the camera projection matrix via scripting is the way to go.

https://user-images.githubusercontent.com/1711423/165708201-8fdc03c3-a661-46a7-86e6-f65a945b2476.mp4

This script has no place here, but just for reference.

```csharp
using System;
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public class NewBehaviourScript : MonoBehaviour
{
    public float v_angle = 10;
    public float h_angle = 10;
    public Matrix4x4 originalProjection;
    Camera cam;

    void Start()
    {
        cam = GetComponent<Camera>();
        originalProjection = cam.projectionMatrix;
    }

    void Update()
    {
        Matrix4x4 p = originalProjection;
        Matrix4x4 shear_matrix = Matrix4x4.identity;
        shear_matrix.m02 = -Mathf.Cos(h_angle * Mathf.Deg2Rad);
        shear_matrix.m12 = -Mathf.Sin(v_angle * Mathf.Deg2Rad);
        cam.projectionMatrix = originalProjection * shear_matrix;
    }
}
```

**Comment by Astrono2** [comment]
> Wow, okay. I sat down to get to work on this and the rendering system has gotten so much more complex since Godot 3. The handling and passing around of `CameraMatrix` objects is all over the place. This will take an enormous amount of reading code and drawing things out to even grasp what files are involved. I don't think I'll get it done by the release of Godot 4.
> 
> Sadly what was a non-trivial but doable change when I proposed it has grown in scope drastically. I think it would be best to close this with @qaptoR's eventual pull request. I still think storing a `Camera3D`'s matrix as a property and using that for rendering is a good idea, but it's no longer a sensible solution to anyone's needs. It would be more of a structural change to allow for easier customization and more user control.
> 
> I will open another issue once I have a better idea of what storing a `Camera3D`'s matrix and using it for rendering even means in the current rendering pipeline.

**Update**:
Ok, so when I posted that comment I had seen the changes to the rendering system and I got scared due to the increased complexity :sweat_smile:. Now that I've finished school, I decided to come back to this project and realized it's not as bad as I thought, as the matrix parameters (fov, near, far, etc.) get turned into a CameraMatrix fairly early on, as the camera's data switches to using the actual matrix at the [`RendererSceneRender::CameraData`](https://github.com/godotengine/godot/blob/1b2992799b324479b3fba9e05ae6226a46cb4143/servers/rendering/renderer_scene_render.h#L242) stage. This means that changes are mainly just necessary in the steps previous to the configuration of the next frame's CameraData.

So what I'm thinking is the following: 

- Change [`RendererSceneCull::Camera`](https://github.com/godotengine/godot/blob/1b2992799b324479b3fba9e05ae6226a46cb4143/servers/rendering/renderer_scene_cull.h#L65) to use just a CameraMatrix for the projection
- Add a `camera_set_matrix` method to `RendererSceneCull` and corresponding `set_matrix` methods to `CameraMatrix` and `Camera3D`
- Add a new case to every switch statement that checks the projection type

My first ideas were unnecessarily complex and required very invasive changes to solve hypothetical future problems. I think this would be the simplest possible solution, achieving the desired feature with minimal changes. I will start working on this and there will be a PR soon hopefully.

**Comment by lyuma** [comment]
Don't forget that a camera may have multiple matrices, for example in XR.

Remember that the culling logic will probably need to take into account the original clipping plane.

the other thing to deal with is oblique matrices may mess with code that tries to convert depth to a world space position. Shadows and other code may need to use an adjustment

What if you could assign an oblique plane to a Camera3D, rather than making it a separate projection mode?

finally, this code by qaptoR is interesting and might be good enough on PC: https://github.com/qaptoR/godot/commit/c12482d9d1a4dc326ca0bd662cf7d05027107d77
it certainly seems like an easier approach than defining custom projection matrices.
That said... Doing `discard;` on Mobile (tile renderers) is expensive, so other options may need to he considered there.

**Comment by Astrono2** [comment]
The oblique plane option is good for portals and is what @qaptoR did. As other people commented here, portals aren't the only use case for custom projection matrices.

For the depth to world space conversion (thanks for the link btw) I think I'll make a non-standard matrix flag and only use the expensive calculation if so. That way my changes don't impact performance for people who aren't using them.

**Comment by Astrono2** [comment]
Ok so storing the matrix seems to be a bad idea because of aspect ratios. Updating the aspect ratio of an existing projection matrix before using it is as messy as generating it when needed, since we would still need a switch statement checking the type (which is what I wanted to avoid). So I'm just going to add the custom projection mode, which will also result in a much smaller PR.

**Comment by clayjohn** [comment]
We discusssed this in today's proposal review meeting and consensus was that we should have a rendering team meeting to discuss further. There are a few key considerations here:
1. This change may add significant complexity to core parts of the engine, so we need to tread carefully
2. We need to decide whether exposing CameraMatrix is worthwhile vs exposing a custom projection type in camera
3. We need to determine if there is a common enough use case that isnt covered by some combination of the existing tools (e.g. using shaders to modify projection matrix, or using frustum mode, or orthogonal mode)

I will start scheduling a meeting, hpoefully we can find a time that works for most of the rendering team and people who are actively interested in this proposal.

**Comment by Zireael07** [comment]
@clayjohn: Can we get a label that says 'needs further review' or something, similar to the 'approved' label?

**Comment by Astrono2** [comment]
> We discusssed this in today's proposal review meeting and consensus was that we should have a rendering team meeting to discuss further. There are a few key considerations here:
> 
>     1. This change may add significant complexity to core parts of the engine, so we need to tread carefully
> 
>     2. We need to decide whether exposing CameraMatrix is worthwhile vs exposing a custom projection type in camera
> 
>     3. We need to determine if there is a common enough use case that isnt covered by some combination of the existing tools (e.g. using shaders to modify projection matrix, or using frustum mode, or orthogonal mode)
> 
> 
> I will start scheduling a meeting, hpoefully we can find a time that works for most of the rendering team and people who are actively interested in this proposal.

For the implementation I'm working on I decided to not expose CameraMatrix and use a custom projection type with a Transform3D and Quaternion. I also noticed that storing the CameraMatrix is a big change in complexity whereas just adding a new projection type leaves the engine code almost as is save a few switch statements and surely some depth and xr stuff.

**Comment by clayjohn** [comment]
Hi everyone, we are planning on discussing this in more depth at a rendering meeting tomorrow at 6:00 am UTC, this time was chosen to accommodate the schedules of contributors in Europe, North America, and Australia. Please feel free to join if you are interested. 

A link will be posted in the #Rendering Channel of the [Godot contributors chat](https://chat.godotengine.org/home) shortly before the meeting starts.

**Comment by Astrono2** [comment]
Sadly I'm from Buenos Aires (GMT-3) so I won't be able to join. Please post here the results of the meeting so I can adapt my (still in early development) PR. Here is my contribution to the discussion:

I think the best way to implement this would be to add CameraMatrix to Variant as I said before (exposing it to the user in any other way makes no sense, all the other math types are exposed like that) or add fixed length typed arrays to be able to pass a `float[4][4]` for the custom projection mode. Passing a Transform3D and Quaternion (or Plane) as I'm currently doing feels like a hack and not suitable for engine code. Maybe the interface for custom projection modes could only be available in C++ for plugins (where we can use `float[4][4]`) and we can leave the botched ways to represent a CameraMatrix in GDScript outside of the engine.

Sorry for not being there tomorrow to explain myself better and participate in the discussion.

**Comment by clayjohn** [comment]
We had a lengthy discussion about this proposal at the rendering meeting today and we think we have reached consensus on the following:
1. Exposing CameraMatrix directly/allowing users to fully specify custom matrices is not worth the added complexity to the renderer. While exposing the matrix itself would not take too much effort, it would not be usable by users as significant changes need to be made to how the renderer handles things like culling, and post-process effects. Godot currently doesn't offer a high degree of control over the core rendering processes, but users would need such control to make their custom projection modes functional.
2. The primary use-case for custom projection matrices appears to be for making portals or mirrors. For portals, it should be enough to expose an oblique projection mode (as is done in https://github.com/godotengine/godot/pull/56664). For mirrors, it should be possible to implement a basic mirror using oblique projection but it will likely not be efficient enough.
3. To properly support high efficiency rendering of portals or mirrors, we need to expose control over the stencil buffer. We plan to do so eventually, but it isn't a high priority right now. 
4. To properly support high quality mirrors, we would prefer to instead implement planar reflections as a built-in engine feature. Doing so allows us to expose high quality reflections for significantly less cost, and in a way that is easy to use for users. 

The above leaves us with two next steps:
1. https://github.com/godotengine/godot/pull/56664 needs to be reviewed to ensure that it is ready for inclusion in to the engine (from a brief look it appears to be well done and can likely be merged without significant changes)
2. We are open to proposals for PlanarReflections so we can discuss implementation details and settle on a way of exposing them that would be useful for a broad subset of users

**Comment by stefano-elysium** [comment]
Hi John, what about the use case for oblique projections (cavalier, etc)? Sadly, I wasn't able to join the discussion.

**Comment by clayjohn** [comment]
@stefano-elysium I imagine it should be possible with the oblique projection mode. If not in the current PR, I'm sure it could be added.

**Comment by Astrono2** [comment]
I think [oblique projections](https://en.wikipedia.org/wiki/Oblique_projection) (cavalier, etc) and [oblique view frustums](https://www.terathon.com/lengyel/Lengyel-Oblique.pdf) (portals and mirrors) are different projections that barely have anything to do with each other.

I'm kinda sad that this feature was rejected but the reasons make a lot of sense. Should I close the issue?

**Comment by lyuma** [comment]
I believe the path forward will adequately address the portal and mirror usecase which was the vast majority of requestors here.

I would say... if you need something other than an oblique frustum (for portals, mirrors), please open a new, *specific* proposal for the type of camera projection needed and why, screenshots and so on

**Comment by elvisish** [comment]
As far as I can tell, this is the only way that the Heretic/Hexen/ROTT y-shearing camera look effect would be possible using a regular perspective camera?

**Comment by Calinou** [comment]
> As far as I can tell, this is the only way that the Heretic/Hexen/ROTT y-shearing camera look effect would be possible using a regular perspective camera?

No, that can already be done with the Frustum projection mode. You'll need to adjust the frustum's near and far values manually to match the desired field of view, and use frustum offset to "look" up and down (instead of rotating the camera up and down).

**Comment by elvisish** [comment]
> > As far as I can tell, this is the only way that the Heretic/Hexen/ROTT y-shearing camera look effect would be possible using a regular perspective camera?
> 
> 
> 
> No, that can already be done with the Frustum projection mode. You'll need to adjust the Camera's frustum **Size** property to match the desired field of view. A frustum size of 0.07 matches a perspective FOV 70 with the default Near and Far values in 3.x (as the FOV depends on near and far when the projection mode is Frustum).
> 
> 
> 
> Then use frustum offset to "look" up and down (instead of rotating the camera up and down).
> 
> 
> 
> Example project in 3.x: [test_y_shearing_3.x.zip](https://github.com/godotengine/godot-proposals/files/9248918/test_y_shearing_3.x.zip)
> 
> 
> 
> See https://github.com/godotengine/godot/issues/63864 and https://github.com/godotengine/godot/issues/63863 for bugs found while using this frustum mode in 3.x and 4.0.

Thanks, that's very helpful! Will this break ’project_ray_origin’ and ’project_ray_normal’ or will they just function as perspective camera? (I'd prefer if frustum shift doesn't affect the ray origin and normal at all)

**Comment by Calinou** [comment]
> Will this break ’project_ray_origin’ and ’project_ray_normal’ or will they just function as perspective camera?

I don't know if these take frustum offset into account, but my guess is that they will act as if you weren't looking up/down in the first place. You'll have to do this manually.

**Comment by elvisish** [comment]
> > As far as I can tell, this is the only way that the Heretic/Hexen/ROTT y-shearing camera look effect would be possible using a regular perspective camera?
> 
> No, that can already be done with the Frustum projection mode. You'll need to adjust the Camera's frustum **Size** property to match the desired field of view. A frustum size of 0.07 matches a perspective FOV 70 with the default Near and Far values in 3.x (as the FOV depends on near and far when the projection mode is Frustum).
> 
> Then use frustum offset to "look" up and down (instead of rotating the camera up and down).
> 
> Example project in 3.x: [test_y_shearing_3.x.zip](https://github.com/godotengine/godot-proposals/files/9248918/test_y_shearing_3.x.zip)
> 
> See [godotengine/godot#63864](https://github.com/godotengine/godot/issues/63864) and [godotengine/godot#63863](https://github.com/godotengine/godot/issues/63863) for bugs found while using this frustum mode in 3.x and 4.0.

I set up a camera with frustum mode to match my perspective camera, it had an FOV of 50 so I managed to get close enough with 0.009 (although 0.0095 would have been most accurate if I'd been able to adjust it that precisely) and while the editor preview works, it just shows a black screen the color of my sky when I run the game, for some reason?

**Comment by elvisish** [comment]
> > > As far as I can tell, this is the only way that the Heretic/Hexen/ROTT y-shearing camera look effect would be possible using a regular perspective camera?
> > 
> > 
> > No, that can already be done with the Frustum projection mode. You'll need to adjust the Camera's frustum **Size** property to match the desired field of view. A frustum size of 0.07 matches a perspective FOV 70 with the default Near and Far values in 3.x (as the FOV depends on near and far when the projection mode is Frustum).
> > Then use frustum offset to "look" up and down (instead of rotating the camera up and down).
> > Example project in 3.x: [test_y_shearing_3.x.zip](https://github.com/godotengine/godot-proposals/files/9248918/test_y_shearing_3.x.zip)
> > See [godotengine/godot#63864](https://github.com/godotengine/godot/issues/63864) and [godotengine/godot#63863](https://github.com/godotengine/godot/issues/63863) for bugs found while using this frustum mode in 3.x and 4.0.
> 
> Thanks, that's very helpful! Will this break ’project_ray_origin’ and ’project_ray_normal’ or will they just function as perspective camera? (I'd prefer if frustum shift doesn't affect the ray origin and normal at all)

Just to confirm, this works, thanks! (at first it was broken because I was setting the offset too far) I've got the size set at 0.009 to match as closely my original 50 FOV perspective camera, but 0.0095 would have been closer, is there a reason it can't be set this low? (0.0095 just reverts to 0.01).

**Comment by Calinou** [comment]
> I've got the size set at 0.009 to match as closely my original 50 FOV perspective camera, but 0.0095 would have been closer, is there a reason it can't be set this low? (0.0095 just reverts to 0.01).

The `size` property hint only allows adjustments by increments of 0.01, but you should be able to adjust Near and Far to control the field of view more precisely (since it affects FOV in Frustum mode).

**Comment by elvisish** [comment]
> 

Yep, I'll adjust those til it matches. Not to derail this any further, but is there a reason the hint doesn't allow for increments of 0.001? I wonder if it would be worth opening a discussion for increasing the hint size.

**Comment by Calinou** [comment]
> Not to derail this any further, but is there a reason the hint doesn't allow for increments of 0.001?

Done: https://github.com/godotengine/godot/pull/63890

**Comment by elvisish** [comment]
> > Not to derail this any further, but is there a reason the hint doesn't allow for increments of 0.001?
> 
> 
> 
> Done for 4.0: https://github.com/godotengine/godot/pull/63890
> 
> 
> 
> This should already be the case in 3.5 as well (but not 3.4).

Thanks! I'm actually on 3.5rc7 so not sure why it's not that fine.

**Comment by stefano-elysium** [comment]
@ghsoares that is what I require :)

**Comment by qaptoR** [comment]
The original `3.4` PR has now been updated to target both `3.x` and `master`.

**Comment by Calinou** [comment]
> Frustrum culling is calculated before the vertex pass (actually, it's with the AABBs, which never get transformed) meaning objects disappear randomly.

You can work this around by increasing **Extra Cull Margin** on your meshes/Sprite3Ds.

**Comment by ghsoares** [comment]
@mrpedrobraga I think you want the same as I did [here](https://github.com/godotengine/godot-proposals/issues/2713#issuecomment-1215138196), if so, take a look at a custom camera projection implementation I made: https://github.com/ghsoares/godot/tree/camera_custom_projection

You may want a skewed camera projection, in the z axis relative to the camera, so the vertices more far away from the camera will be skewed to give the visual you want. So a custom projection like this should work for you:
![image](https://user-images.githubusercontent.com/43936806/192565914-334953eb-3480-438e-b389-602b3dfadf97.png)
The implementation solves the two problems you were having, as the projection matrix used by the renderer is transformed by this custom projection matrix, the lightning will look the same as a regular camera and the frustum culling will work correctly with the skewed vertices.

**Comment by mrpedrobraga** [comment]
@ghsoares Yep, precisely this. This is the kind of feature I expected the engine to already have: when I set `rotation_mode` to basis I thought it'd allow me to skew the camera at least.

Alas, the implementation here is for 3.x, and I don't have the experience with the engine's internals, writing code in the engine, much less porting from 3.x, to 4.x; would you be free/willing to talk about this somewhere? And in the end you'll have a 4.x implementation if you ever need it.

**Comment by ghsoares** [comment]
@mrpedrobraga I see, I could implement in 3.x because my machine have support for OpenGL only, and 4.x is really buggy with opengl rendering command option, so I couldn't implement and test this feature for 4.x. But I think the idea would be the same, if you can locate some part of the code that creates a camera/projection matrix to use in the scene shaders in the rendering server, you could try to multiply this matrix by the custom projection matrix.

Other option I think would work too is to allow to Camera node to have non-orthonormalized transform, AFAIK the code forces the orthonormalization of the Camera transform before passing to the scene shaders, so if there was a way to allow any kind of transform with a option in the camera, it would work exactly like the `custom_projection` option.

**Comment by jamesblight** [comment]
> I think [oblique projections](https://en.wikipedia.org/wiki/Oblique_projection) (cavalier, etc) and [oblique view frustums](https://www.terathon.com/lengyel/Lengyel-Oblique.pdf) (portals and mirrors) are different projections that barely have anything to do with each other.
> 
> I'm kinda sad that this feature was rejected but the reasons make a lot of sense. Should I close the issue?

Just want to highlight this comment as I think the title of the proposed PR is misleading https://github.com/godotengine/godot/pull/56664. It doesn't enable actual oblique projection and I wouldn't call it an oblique camera type. It solves the portal usecase but doesn't enable oblique projection types like military, cavalier and cabinet.

**Comment by XavierH622** [comment]
I am working on a game that has portals as a Key game mechanic. I'd love to use Godot for this game but i cant until this feature is added. Just wanted to show my support.

**Comment by Calinou** [comment]
@XavierH622 lease don't bump issues without contributing significant new information. Use the :+1: reaction button on the first post instead.

**Comment by ghsoares** [comment]
How about allowing non-orthonormalized transform for the camera? It could be a new property for the camera resource (in the RenderingServer) and the camera node with default value to true, but the developer can disable it to allow skewed camera projections, for example. I don't know if it would help for every case of oblique projection, but it would help a lot for skewed orthonormalized camera projection (basically what I and @mrpedrobraga wanted to do) for example.

I can create a PR for this feature myself, as Godot 4 reached stable release some time now, I can work with the core code more confidently, it seems to be pretty straightforward to do as I already something simillar in 3.x ([see this fork commit I did for 3.x](https://github.com/ghsoares/godot/commit/a37a839597acf7cf592b8a38b02140e699552761))

**Comment by DxUr** [comment]
I done that in Godot 4

![Screenshot from 2023-08-19 09-38-27](https://github.com/godotengine/godot-proposals/assets/137767403/ec78597c-0173-4993-892e-2e302b9035b8)

**Comment by AThousandShips** [comment]
Feel free to open a PR to implement this feature!

**Comment by Power-Jake** [comment]
> I done that in Godot 4
> 
> ![Screenshot from 2023-08-19 09-38-27](https://user-images.githubusercontent.com/137767403/261788977-ec78597c-0173-4993-892e-2e302b9035b8.png)

Hi, have you shared the code for this change anywhere? I would love to see it.

**Comment by DxUr** [comment]
> > I done that in Godot 4
> > ![Screenshot from 2023-08-19 09-38-27](https://user-images.githubusercontent.com/137767403/261788977-ec78597c-0173-4993-892e-2e302b9035b8.png)
> 
> Hi, have you shared the code for this change anywhere? I would love to see it.

Actually no but I will port this to 4.2 because it's done in 4.1 then I will make a pull request

**Comment by curtishimel** [comment]
Did this make it into the 4.2 beta that's currently in testing?

**Comment by DxUr** [comment]
> DxUr, did that custom projections PR make it into the 4.2 beta that's currently in testing?

This is my implementation, idk if it added to 4.2 then I'll make a pull request later. I'm so lazy to do that.

**Comment by PitouGames** [comment]
I'm very interested in this feature, it will be really usefull to create application for [VR CAVE](https://en.wikipedia.org/wiki/Cave_automatic_virtual_environment).

I can't find any PR made by @DxUr. There is also no trace of custom projection in [4.2-beta1](https://github.com/godotengine/godot-builds/releases/tag/4.2-beta1)...

**Comment by mathiasmellemstuen** [comment]
I'm also interested in this feature when I need this for doing oblique projection planes. Currently working on a project with portals and non-euclidean worlds where modifying the projection matrix is a must.

**Comment by DxUr** [comment]
@mathiasmellemstue
@PitouGames 
the implementation on Godot 4.2 I think is different, this is why you can not find any PR by me.
then if you  need it  a lot tell me, I'll make it accessible from code NOT GUI, new people can break the projection and say GODOT is a bad blabla.

**Comment by mathiasmellemstuen** [comment]
> @mathiasmellemstuen @PitouGames the implementation on Godot 4.2 I think is different, this is why you can not find any PR by me. then if you need it a lot tell me, I'll make it accessible from code NOT GUI, new people can break the projection and say GODOT is a bad blabla.

I totally agree with this. Making the projection matrix available only from code is probably the right way based on the reasoning you stated. Not sure how many will benefit from setting the coefficients manually in editor anyways.

Personally I would really appreciate if you go through with this.

**Comment by DxUr** [comment]
@mathiasmellemstuen ok, I'll do it, I hope they accept it for 4.2.

**Comment by Gamemap** [comment]
Would it be possible with the Camera Matrix to have an off-center perspective for a 3d camera?
<details><summary>Here is an example</summary>

![off center](https://github.com/godotengine/godot-proposals/assets/71942164/d97ea25e-29b4-4605-a82e-b9b47d269504)
</details>

**Comment by Calinou** [comment]
> @mathiasmellemstuen ok, I'll do it, I hope they accept it for 4.2.

4.2 is in feature freeze, so any new features will have to target 4.3 at the earliest.

**Comment by DxUr** [comment]
@mathiasmellemstuen 
@PitouGames 
I do a PR for custom projection matrix implementation.
see https://github.com/godotengine/godot/pull/84454

**Comment by DxUr** [comment]
Can any one tell the team to merge it, I really need it in my projects and compiling from source is waste of time.

**Comment by devalexx** [comment]
could we please merge the PR? from the code it's clear that the code doesn't break existing functionality, but bring the new one, which was requested already 3y ago and exists in unity for quite long time. or at least prepend '[experimental]' label on the dropdown

**Comment by Calinou** [comment]
> could we please merge the PR? from the code it's clear that the code doesn't break existing functionality, but bring the new one, which was requested already 3y ago and exists in unity for quite long time. or at least prepend '[experimental]' label on the dropdown

Some issues were reported with that PR: https://github.com/godotengine/godot/pull/84454#issuecomment-1875073902

Also, it still needs a review from a rendering maintainer before it can be merged.

**Comment by amonroejj** [comment]
My use case for custom projections in Camera3D is to orthogonally skew the whole world to give walls the appearance of height even when the camera is pointing straight down (mockup faked in Blender by moving the verts of the wall tops). This seems like a common use case for various kinds of 2.5 top down games.

![image](https://github.com/godotengine/godot-proposals/assets/2057161/0ed0ca4f-9f8c-405d-a8f8-5866d624eb69)

Related technique is discussed on this page: https://manabreak.github.io/devlog/satos/3d/2019/08/16/3d-to-2d.html

**Comment by pineapplemachine** [comment]
I would really appreciate the option to experiment with different projections, e.g. spherical projection, without having to do weird workarounds in spatial shaders. What Godot projects are currently missing in 3D fidelity compared to other major engines can be made up for with art direction, but access to the projection matrix is going to be an important tool for creating a distinctive art direction. Is there anything those of us hoping for this feature can do to help move this forward?

**Comment by pineapplemachine** [comment]
Oh yes, many of the most interesting effects require more than _just_ a custom camera projection matrix. But this is certainly one step along the way, and a helpful tool to have in any case.

**Comment by pyrareae** [comment]
This would be a greatly valuable feature for my use case. Lack of custom projection is a holdup for moving a concept I've been working on from Unity to Godot.

**Comment by periodyctom** [comment]
This would be useful for me as well.

**Comment by periodyctom** [comment]
Now that 4.3 is out, how is it looking for getting this into 4.4?

**Comment by majikayogames** [comment]
I'm still using https://github.com/godotengine/godot/pull/85529 as this is giving me culling issues with mirrors:
https://www.youtube.com/watch?v=mB1SFQdHEYY

Both seem to have cull issues, I think #85529 still breaks in the case of lights sometimes which could normally be discounted if the are occluded. But I can work around that by adding some cull margins on things.

**Comment by JNSStudios** [comment]
I'm new to this engine, and I wanted to basically recreate the effect described in this Stack Overflow question for Unity: https://gamedev.stackexchange.com/questions/191696/how-can-i-set-up-a-3d-camera-for-a-top-down-game-like-enter-the-gungeon/206850#206850 

And of course the only way to do it is to directly edit the camera projection matrix, which you can't do. And of course I find this out the day 4.4 launches without this feature. I hope they allow for this feature in the future.

**Comment by ultrasuperpingu** [comment]
I'd like to use Godot with a CAVE (cave automatic virtual environment) system. I'm just looking for information for now so I have to check but I'm not sure it's possible with the current PROJECTION_FRUSTUM mode... It would be much easier to have the ability to set the projection matrix manually. I don't really understand why this setter is not provided ? This feature is present in all other 3d engine I worked with (Unity, UE, Ogre, Panda3D, ...) :(
I see another use case I already used this feature: make a smooth transition between orthographic and perspective projection :)

**Comment by lyuma** [comment]
Yes, the code at https://github.com/V-Sekai/godot_planar_reflection can be used for building a CAVE style system using Godot's frustum camera. Here is an example showing how the frustum planar reflection actually works using my addon:

https://github.com/user-attachments/assets/8ed061d8-5f90-4888-bc84-7f6fd7f807c

**Comment by ultrasuperpingu** [comment]
@lyuma Nice. I'll check your code later. But even if it is possible to do it with the current api, I'm still thinking it would be way easier to just get a setter to the camera projection matrix :)
Thanks for the sharing 👍

**Comment by ultrasuperpingu** [comment]
The PROJECTION_FRUSTUM mode does the job to work with a CAVE system :) And I saw an other proposal (#11436 ) summarize the use cases that would be solved by proposing an deeper acces to the projection matrix. I guess the main issue to provide a setter to the Camera Projection Matrix is that it would be more difficult to provide an efficent culling. Am I right ?

**Comment by mrpedrobraga** [comment]
That plus the whole "pretty much the whole engine assumes things about the Camera's matrix."

For example, when I tried patching Godot myself to allow arbitrary matrices, I recall it breaking lighting, shadows, mouse ray casting, AABBs...

**Comment by ultrasuperpingu** [comment]
@mrpedrobraga Ok, I don't really understand why tho...

**Comment by xne** [comment]
There's some discussion over in #11436 on how to make the camera matrix accessible without breaking the engine's assumptions

**Comment by AdAstraSaturn** [comment]
is there a way to achieve oblique projection with this? like enter the gungeon

**Comment by mrpedrobraga** [comment]
> is there a way to achieve oblique projection with this? like enter the gungeon

Yes! Simply set the matrix such that the Z vector has a Y component.

![Oblique Projection](https://media.discordapp.net/attachments/900127562905296956/1367234359215784127/image.png?ex=68152939&is=6813d7b9&hm=5f54a3d6fbb9fa3d1c00e4379e3911da9e184f8e8e0015c62d20cb568c69e92d&=&format=webp&quality=lossless&width=885&height=921)

Here's it applied to Kenney's 3D Platformer Starter Kit.

![Preview Video](https://www.youtube.com/watch?v=_MopWSkbHuE)

Here, the Z basis vector has the same length as the Y basis vector... so squares in the world in the XZ plane still look like squares.

![](https://media.discordapp.net/attachments/900127562905296956/1367234658462597185/image.png?ex=68152980&is=6813d800&hm=e620481e206b44ac3ee77aa3c871e7c0931aea577639a6764175c476d74bcc2b&=&format=webp&quality=lossless&width=885&height=921)

You can even make the Z axis have a X component too, though why you'd do it is beyond me.

![](https://media.discordapp.net/attachments/900127562905296956/1367234865933713511/image.png?ex=681529b2&is=6813d832&hm=78b83428825b2c901584bde67464181cf7c6b6d58f452298fb2509bbcf114503&=&format=webp&quality=lossless&width=885&height=921)
